src/core/view.ts

```diff
@@ -50,12 +50,15 @@ export abstract class View<E extends Element, S extends Snapshot<E> = Snapshot<E
     }
 
     const { isPreview, shouldRender, newSnapshot: snapshot } = renderer
+
     if (shouldRender) {
       try {
         this.renderer = renderer
         this.prepareToRenderSnapshot(renderer)
-        this.delegate.viewWillRenderSnapshot(snapshot, isPreview)
-        await this.renderSnapshot(renderer)
+        const event = this.delegate.viewWillRenderSnapshot(snapshot, isPreview)
+        if (!event.defaultPrevented) {
+          await this.renderSnapshot(renderer)
```

Comment:
Snapshot rendering involves more than setting `innerHTML` (e.g. permanent element transfer).

I wonder if _not_ preventing the default, and instead providing some form of "renderer" interface as part of the `event.detail` payload, then allow the event listeners to provide their own (via `event.detail.render = ...`).

Reply:
You are right. I didn't consider that.

I was exploring another route as well. It is possible to overload an `Element`'s `replaceChild` function. Either by monkey-patching it:
```typescript
document.documentElement.replaceChild = <T extends Node>(newChild: Node, oldChild: T) {
  const clone = oldChild.cloneNode(true) as T;
  morphdom(oldChild, newChild);
  return clone;
};
```
Or by extending the elements with web components. Its usage could look like this:
```html
  <body is="morphable">
```
Turbo Frames could be extended in a similar fashion (`class MorphableTurboFrame extends TurboFrame { ...`). The good thing about this solution would be that it requires basically no change to turbo drive. Only the [cloning of the body before caching it](https://github.com/hotwired/turbo/pull/218/files#diff-1e496dfcade128e45d250f8728df33466d2fd2b35303d626be406cb4f92083a5R17) would be needed for mutating renderers. I'm still a bit on the fence what the cleanest solution would be.

Reply:
Well I guess it is up to @sstephenson to decide how the API should look like. So some input about the general direction would be much appreciated.